Sort more data types

notlee opened the pull request:
Add the ability to sort new types of data.

**Reference**
FT - Types of cell content:
https://gist.github.com/cfq/5c6349443735bb10b3470ae75a2543c1
Other date references:
- "22nd Aug 2018 3:17pm" (Email Platform)
- "16-07-2018, 14:20" (gdpr-sar-hub)

**Todo**
- [x] Links
- [x] Integers
- [x] Floats
- [x] Formatted integers (43,234)
- [x] Formatted floats (54,234.55)
- [x] Abbreviated integers (15bn)
- [x] Percentages
- [x] FT Style Partial dates (May 4)
- [x] FT style abbreviated partial dates (Feb 6)
- [x] FT Style date ranges (Jun 6‑7) -- *First date in range used*
- [x] Empty cells
- [x] Text/numbers/links with asterisk or links to footnotes
- [x] Images -- *based on alt*
- [x] Email icon as ```<i class="list-view__icon o-icons-icon o-icons-icon--mail"><a href="mailto:andrew.rininsland@ft.com" title="Email Ændrew Rininsland @ andrew.rininsland@ft.com"></a>``` -- *will sort based on `aria-label` or `title` if the cell has no other content*
- [x] (Integer|Float|Date) range -- *currently uses the first value in the range*
- [x] Text in number columns (New) -- *Sorted before numerical values (ASC), after (DES).*
- [x] Cells containing null values ('-', 'N/A') -- *Sorted before numerical or string values (ASC), after (DES)*

**Assumed out of scope (check)**
- ✔ as ```<i class="list-view__icon o-icons-icon o-icons-icon--tick"></i>``` or ```⨯``` -- *do not want to make assumptions about classes, however will use an aria-label or title if provided*
- Stars (★★★★★★) -- *sort value can be provided via data attribute*
- Bar charts -- *sort value can be provided via data attribute*
- Animated bar charts
- Segmented bar charts
- Dates - FT style dates are supported, otherwise define a custom sort filter per the documentation
- Formatted dates (25 October 2005) - FT style dates are supported, otherwise define a custom sort filter per the documentation

cfq commented:
Excellent work @notlee!

notlee commented:
Thanks @cfq -- missing you!

gvonkoss commented on src/js/sort-formatter.js:
```diff
@@ -0,0 +1,298 @@
+}
+
+/**
+ * Returns the text with non-number charactors removed (e.g. currency symbols).
```
charactors -> characters (on the line below, also)

gvonkoss commented on README.md:
```diff
@@ -239,13 +278,92 @@ If you are wanting to sort by a custom pattern, you can apply the sorting values
+
+##### Custom sort (imperative)
+
+Rather than specify `data-o-table-sort-value` [declaritively](#custom-sort-declarative), a formatter function may be provided client-side to generate sort values for a given data type.
```
declaritively -> declaratively

gvonkoss commented on README.md:
````diff
@@ -239,13 +278,92 @@ If you are wanting to sort by a custom pattern, you can apply the sorting values
+</table>
+```
+
+Note: `data-o-table-order` has been deprecated in favour of `data-o-table-sort-value` to specify a sort value.
````
If this release is going to be a breaking change, we probably don't need to warn about deprecations :v:

notlee replied:
Good shout. As we're likely to make other significant changes to `o-table` soon I'm going to avoid a breaking change if I can without having to do any awkward work arounds. If this does become a breaking change I'll make a note to remove deprecations :)

JakeChampion commented on src/js/oTable.js:
```diff
@@ -131,219 +23,347 @@ function descendingSort(...args) {
+
+class OTable {
+	/**
+	 * Initialises an o-table components inside the element passed as the first parameter
```
> Initialises an o-table _component_ inside the element passed as the first parameter

Pedantic: The constructor first checks to see if the element itself is an o-table component, if it is not it then looks inside the element to find the first instance of an o-table and constructs that.

JakeChampion commented on src/js/oTable.js:
```diff
@@ -131,219 +23,347 @@ function descendingSort(...args) {
+				console.warn('o-table: "data-o-table-order" is deprecated, to provide a custom sort value for a table cell use "data-o-table-sort-value" instead.');
+				cellsWithOrder.forEach(cell => {
+					if (cell.getAttribute('data-o-table-order') !== null) {
+						cell.setAttribute('data-o-table-sort-value', cell.getAttribute('data-o-table-order'));
```
Should we also remove the deprecated attribute after copying it's value over to the new attribute name?

notlee replied:
As we can't know if implementers are relying on this attribute for some reason I think not, as there's no obvious benefit.

JakeChampion commented on src/js/oTable.js:
```diff
@@ -131,219 +23,347 @@ function descendingSort(...args) {
-			return ascendingSort(aCol, bCol, isNumericValue, intlCollator);
-		} else {
-			return descendingSort(aCol, bCol, isNumericValue, intlCollator);
+			this.tableHeaders = Array.from(this.rootEl.querySelectorAll('thead th'));
```
I _think_ this will also match `th` elements which are not within a `thead` due to a quirk in querySelectorAll where by default it only verifies that the last element in the selector is within the search scope.

If we want to ensure we only return `th` elements which are within a `thead` we need to use `this.rootEl.querySelectorAll(':scope thead th')` however, that is not supported in IE or Edge and requires a polyfill being added to polyfill.io 😩

notlee replied:
I shall: 

```
const tHead = this.rootEl.querySelectorAll('thead')
this.tableHeaders = tHead.querySelectorAll('th')
```

JakeChampion commented on src/js/oTable.js:
```diff
@@ -131,219 +23,347 @@ function descendingSort(...args) {
+
+				const listener = this._sortByColumn(columnIndex);
+				this.listeners.push(listener);
+				th.addEventListener('click', listener);
```
We should look into using [dom-delegate](https://www.npmjs.com/package/dom-delegate) again, I'm not sure why we stopped using it 🤔

notlee replied:
If the next release ends up being a breaking I'll add it
https://github.com/Financial-Times/o-table/issues/103

JakeChampion commented on src/js/oTable.js:
```diff
@@ -131,219 +23,347 @@ function descendingSort(...args) {
+				th.addEventListener('keydown', (event) => {
+					const ENTER = 13;
+					const SPACE = 32;
+					if ('code' in event) {
```
Should we use [`key`](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/key) instead of `code`?

notlee replied:
I'm not sure that matters here? For ENTER and SPACE, we don't care about modifier keys such as shift?

JakeChampion commented on src/js/oTable.js:
```diff
@@ -131,219 +23,346 @@ function descendingSort(...args) {
-		 * This is called again by the `sorted` method to assure accuracy.
-		 * I.e. if a sort fails previous sort attributes can be restored via the `sorted` method.
-		 */
+		window.requestAnimationFrame(() => {
```
Need to include this in the browser features in origami.json

notlee replied:
Looks like `requestAnimationFrame` is supported in all browsers we support an enhanced experience.

notlee commented:
@JakeChampion as this PR focuses on new sort functionality, and outstanding comments are of existing code which has been moved to use the class syntax, I'll merge this into a release branch and follow up with another PR that refactors and addresses those comments. :)
